openghg/retrieve/icos/_retrieve.py

```diff
@@ -0,0 +1,305 @@
+        dobj = Dobj(dobj_url)
+        # We need to pull the data down as .info (metadata) is populated further on this step
+        dataframe = dobj.get()
+        # This is the metadata
+        dobj_info = dobj.info
+
+        metadata = _extract_metadata(meta=dobj_info, site_metadata=site_metadata)
+
+        # There's some extra metadata available from their API
+        # that isn't currently (2022-04-28) available using ICOS pylib
+        metadata_url = f"{dobj_url}/meta.json"
+        metadata_bytes = download_data(url=metadata_url)
+
+        if metadata_bytes is not None:
+            extra_metadata = json.loads(metadata_bytes)
+
+            to_add: Dict[str, Union[str, List, Dict]] = {}
+
+            try:
+                reference_data = extra_metadata["references"]
+            except KeyError:
+                to_add["citation_string"] = "NA"
+                to_add["licence"] = "NA"
+            else:
+                to_add["citation_string"] = reference_data.get("citationString", "NA")
+                to_add["licence"] = reference_data.get("licence", {}).get("baseLicence", "NA")
+
+            try:
+                instrument_metadata = extra_metadata["specificInfo"]["acquisition"]["instrument"]
+            except KeyError:
+                to_add["instrument"] = "NA"
+                to_add["instrument_data"] = "NA"
+            else:
+                # Do some tidying of the instrument metadata
+                instruments = set()
+                cleaned_instrument_metadata = []
+
+                if not isinstance(instrument_metadata, list):
+                    instrument_metadata = [instrument_metadata]
+
+                for inst in instrument_metadata:
+                    instrument_name = inst["label"]
+                    instruments.add(instrument_name)
+                    uri = inst["uri"]
+
+                    cleaned_instrument_metadata.extend([instrument_name, uri])
+
+                if len(instruments) == 1:
+                    instrument = instruments.pop()
+                else:
+                    instrument = "multiple"
+
+                to_add["instrument"] = instrument
+                to_add["instrument_data"] = cleaned_instrument_metadata
+
+            metadata.update(to_add)
+
+        # Add ICOS in directly here for now
+        metadata["network"] = "ICOS"
+        metadata["data_type"] = "timeseries"
+        metadata["data_source"] = "icoscp"
+
+        dataframe.columns = [x.lower() for x in dataframe.columns]
+        dataframe = dataframe.dropna(axis="index")
+
+        if not dataframe.index.is_monotonic_increasing:
+            dataframe = dataframe.sort_index()
+
+        spec = metadata["species"]
+        inlet = metadata["inlet"]
+
+        rename_cols = {
+            "stdev": spec + " variability",
+            "nbpoints": spec + " number_of_observations",
+        }
+
+        dataframe = dataframe.rename(columns=rename_cols).set_index("timestamp").astype({"flag": str})
```

Comment:
~/openghg/openghg/retrieve/icos/_retrieve.py in retrieve(site, species, sampling_height, start_date, end_date, force_retrieval, data_level)
     52     else:
     53         # We'll also need to check we have current data
---> 54         standardised_data = _retrieve_remote(site=site, species=species, data_level=data_level)
     55 
     56         if standardised_data is None:

~/openghg/openghg/retrieve/icos/_retrieve.py in _retrieve_remote(site, data_level, species, sampling_height)
    217         }
    218 
--> 219         dataframe = dataframe.rename(columns=rename_cols).set_index("timestamp").astype({"flag": str})
    220         dataframe.index.name = "time"
    221         dataframe.index = to_datetime(dataframe.index, format="%Y-%m-%d %H:%M:%S")

Reply:
For the line
```dataframe.rename(columns=rename_cols).set_index("timestamp").astype({"flag": str})```
this looks like "flag" is "userflag" in this particular DataFrame (not sure if this is true more universally?)

Reply:
For the line
```dataframe.rename(columns=rename_cols).set_index("timestamp").astype({"flag": str})```
this looks like "flag" is "userflag" in this particular DataFrame (not sure if this is true more universally?)